lib/bolt_ext/puppetdb_inventory.rb

```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/env ruby
+
+require 'json'
+require 'httpclient'
+require 'optparse'
+require 'yaml'
+
+module Bolt
+  class PuppetDBInventory
+    class Client
+      def self.from_config(config)
+        uri = URI.parse(config['server_urls'].first)
+        uri.port ||= 8081
+
+        cacert = File.expand_path(config['cacert'])
+        if config['token']
+          token = File.read(File.expand_path(config['token'])).chomp
+        end
+        if config['cert'] && config['key']
+          cert = File.expand_path(@cert)
+          key = File.expand_path(@key)
+        end
+
+        new(uri, cacert, token: token, cert: cert, key: key)
+      end
+
+      def initialize(uri, cacert, token: nil, cert: nil, key: nil)
+        @uri = uri
+        @cacert = cacert
+        @token = token
+        @cert = cert
+        @key = key
+      end
+
+      def query_certnames(query)
+        return [] unless query
+
+        body = JSON.generate(query: query)
+
+        response = http_client.post("#{@uri}/pdb/query/v4", body: body, header: headers)
+        if response.code != 200
+          raise "Failed to query PuppetDB: #{response.body}"
+        else
+          results = JSON.parse(response.body)
+          if results.first && !results.first.key?('certname')
+            raise "Query results did not contain a 'certname' field: got #{results.first.keys.join(', ')}"
+          end
+          results.map { |result| result['certname'] }.uniq
+        end
+      end
+
+      def http_client
+        return @http if @http
+        @http = HTTPClient.new
+        @http.ssl_config.set_client_cert_file(@cert, @key)
+        @http.ssl_config.add_trust_ca(@cacert)
+
+        @http
+      end
+
+      def headers
+        headers = { 'Content-Type' => 'application/json' }
+        headers['X-Authentication'] = @token if @token
+        headers
+      end
+    end
+
+    class Config
+      DEFAULT_TOKEN = File.expand_path('~/.puppetlabs/token')
+
+      def initialize(options)
+        @settings = load_config(File.expand_path('~/.puppetlabs/client-tools/puppetdb.conf'))
```

Comment:
I think it'd make sense for this to be configurable as well, since the config could live in a different location.